specutils/manipulation/resample.py

```diff
@@ -0,0 +1,293 @@
+from abc import ABC, abstractmethod
+
+import numpy as np
+from scipy.interpolate import CubicSpline
+from astropy.units import Quantity
+from astropy.nddata import StdDevUncertainty, VarianceUncertainty, InverseVariance
+
+from ..spectra import Spectrum1D
+
+__all__ = ['ResampleBase', 'FluxConservingResample']
+
+
+class ResampleBase(ABC):
+    """
+    Base class for resample classes.  The algorithms and needs for difference
+    resamples will vary quite a bit, so this class is relatively sparse.
+
+    Init paramtere here is not yet hooked up to the rest of the code, but to
+    show how we will want to use it in the future.
+    """
+    def __init__(self, bin_edges='zero_fill'):
+        self.bin_edges = bin_edges
+
+    @abstractmethod
+    def __call__(self, orig_spectrum, fin_lamb):
+        """
+        Return the resulting `~specutils.Spectrum1D` of the resampling.
+        """
+        return NotImplemented
+
+    @abstractmethod
+    def resample1D(self, orig_spectrum, fin_lamb):
+        """
+        Workhorse method that will return the resampled Spectrum1D
+        object.
+        """
+        return NotImplemented
+
+    @staticmethod
+    def _calc_bin_edges(x):
+        """
+        Calculate the bin edge values of an input dispersion axis. Input values
+        are assumed to be the center of the bins.
+
+        todo: this should live in the main spectrum object, but we're still
+        figuring out the details to that implementation, so leaving here
+        for now.
+
+        Parameters
+        ----------
+        x : ndarray
+            The input dispersion axis values.
+
+        Returns
+        -------
+        edges : ndarray
+            Calcualated bin edges, including left and right most bin edges.
+        """
+        inside_edges = (x[1:] + x[:-1]) / 2
+        edges = np.insert(inside_edges, 0, 2 * x[0] - inside_edges[0])
+        edges = np.append(edges, 2 * x[-1] - inside_edges[-1])
+
+        return edges
+
+
+class FluxConservingResample(ResampleBase):
+    """
+    This resample algorithim conserves overall flux (as opposed to flux density).
+    Algorithim based on the equations documented in the following paper:
+    https://ui.adsabs.harvard.edu/abs/2017arXiv170505165C/abstract
+    """
+
+    def __call__(self, orig_spectrum, fin_lamb):
+        """
+        Return the resulting `~specutils.Spectrum1D` of the resampling.
+        """
+        return self.resample1D(orig_spectrum, fin_lamb)
+
+    def _resample_matrix(self, orig_lamb, fin_lamb):
+        """
+        Create a re-sampling matrix to be used in re-sampling spectra in a way
+        that conserves flux. This is adapted from
+        https://ui.adsabs.harvard.edu/abs/2017arXiv170505165C/references,
+        eprint arXiv:1705.05165. This code was heavily influenced by Nick Earl's
```

Comment:
Can you reference either a PR # or a commit on github to cite this just so that there's history?

Reply:
Sure, there's no PR though.  I was referencing his branch off his fork: https://github.com/nmearl/specutils/commit/0ff6ef169b5607a1e059cb7cc9488238191f3c39.  What's the best way to reference that?

Reply:
I think it's find just to give the URL directly to that SHA (e.g. https://github.com/nmearl/specutils/commit/0ff6ef169b5607a1e059cb7cc9488238191f3c39 ) - this is a private method so no need to be too picky about it